### PR TITLE
Selective tests - depends on files changed in the commit.

### PR DESCRIPTION
**Pull request by potiuk**
This is final step of implementing #10507 - selective tests.
Depending on files changed by the incoming commit, only subset of
the tests are exucted. The conditions below are evaluated in the
sequence specified below as well:

* In case of "push" and "schedule" type of events, all tests
  are executed.

* If no important files and folders changed - no tests are executed.
  This is a typical case for doc-only changes.

* If any of the environment files (Dockerfile/setup.py etc.) all
  tests are executed.

* If no "core/other" files are changed, only the relevant types
  of tests are executed:

  * API - if any of the API files/tests changed
  * CLI - if any of the CLI files/tests changed
  * WWW - if any of the WWW files/tests changed
  * Providers - if any of the Providers files/tests changed

* Integration Heisentests, Quarantined, Postgres and MySQL
  runs are always run unless all tests are skipped like in
  case of doc-only changes.

* If "Kubernetes" related files/tests are changed, the
  "Kubernetes" tests with Kind are run. Note that those tests
  are run separately using Host environment and those tests
  are stored in "kubernetes_tests" folder.

* If some of the core/other files change, all tests are run. This
  is calculated by substracting all the files count calculated
  above from the total count of important files.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
I am still testing that one in my private fork (https://github.com/potiuk/airflow/pulls). There are 12 pull requests that (I hope) should run faster or longer, and have more or less checks depending on the files that are changed in the upcoming commit. 

This is final step of implementing #10507  and I think this will be ultimate saver on the job queues that we have in Github Actions - and for most contributors it will significantly speed up the feedback time - if they only change docs + some of the providers for example. I will use the linked 12 PRs from my fork  to make sure that it works as expected before merging but I am happy to get reviews before.

File: scripts/ci/selective_tests.sh [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+    initialization::ga_output kind-versions \
+        "$(initialization::parameters_to_json "${CURRENT_KIND_VERSIONS[@]}")"
+    initialization::ga_output default-kind-version "${KIND_VERSION}"
+
+    initialization::ga_output helm-versions \
+        "$(initialization::parameters_to_json "${CURRENT_HELM_VERSIONS[@]}")"
+    initialization::ga_output default-helm-version "${HELM_VERSION}"
+
+    initialization::ga_output postgres-exclude '[{ "python-version": "3.6" }]'
```
Comment (potiuk) [inline scripts/ci/selective_tests.sh:52]: If we have as many savings as I think we will, we can likely even bring back all python variants for all tests.

**Comment by potiuk** [comment]
Hello everyone - this one is ready for review and it looks like it works as expected:

You can see all the 13 different cases where I tested small changes in some of the combinations: 

https://github.com/potiuk/airflow/pulls

Examples: 

* API changes: https://github.com/potiuk/airflow/pull/113 when there are only API files changed -51 checks are run rather than 110.
* When important files such as setup.py are changed - all tests are run : https://github.com/potiuk/airflow/pull/124
* When you push directly to master (that's how merge works), all tests are run: https://github.com/potiuk/airflow/runs/1238562869?check_suite_focus=true#step:5:452 (I cancelled this one but you can see tht it would run all tests)
* When there is a kubernetes_tests/chart related change it will run Kubernetes tests: https://github.com/potiuk/airflow/pull/121 and https://github.com/potiuk/airflow/pull/122
* When there are providers-only changes (or provider tests) https://github.com/potiuk/airflow/pull/117/  also ~ 60 checks are run (but this time the Providers one are among them)
* When there is a combined change with several directories - it will run all relevant tests: https://github.com/potiuk/airflow/pull/125

**Comment by potiuk** [comment]
Would be great to merge this one for the coming weel  - it will significantly decrease the strain on the GA.

**Review by kaxil — APPROVED** [review]

**Pull request merged**